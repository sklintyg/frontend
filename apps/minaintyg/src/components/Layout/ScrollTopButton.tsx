import { classNames } from '@frontend/components'
import { IDSButton, IDSIconArrow } from '@frontend/ids-react-ts'
import { getTabbableNodes } from 'focus-lock'
import { useEffect, useState } from 'react'

export function ScrollTopButton() {
  const [show, setShow] = useState(false)
  const [isScrolling, updateIsScrolling] = useState(false)

  useEffect(() => {
    let previousScrollPos = 0
    const handleScroll = () => {
      const scrollPos = document.body.getBoundingClientRect().top
      const isScrollInterupted = previousScrollPos > 0 && scrollPos > previousScrollPos

      setShow(scrollPos < -180)
      if (isScrolling && (scrollPos === 0 || isScrollInterupted)) {
        updateIsScrolling(false)
        const mainElement = document.querySelector('main')
        if (!isScrollInterupted && mainElement instanceof HTMLElement) {
          getTabbableNodes([mainElement], new Map())[0].node.focus()
        }
        previousScrollPos = 0
      }
      previousScrollPos = scrollPos
    }

    window.addEventListener('scroll', handleScroll, true)
    return () => {
      window.removeEventListener('scroll', handleScroll, true)
    }
  }, [isScrolling])

  return (
    <div
      className={classNames(
        'sticky bottom-0 h-16 opacity-0 transition-opacity ease-in-out pointer-events-none',
        show && 'opacity-100 pointer-events-auto'
      )}
    >
      <IDSButton
        fab
        role="button"
        className="ids-hide ids-show-1177 absolute bottom-5 right-5 select-none"
        aria-label="Till toppen av sidan"
        onClick={() => {
          updateIsScrolling(true)
          window.scrollTo({
            top: 0,
            left: 0,
            behavior: 'smooth',
          })
        }}
      >
        <IDSIconArrow color="var(--color-main)" width="24" height="24" rotate="-90" />
      </IDSButton>
    </div>
  )
}
