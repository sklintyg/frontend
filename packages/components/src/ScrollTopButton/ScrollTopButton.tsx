import { IDSButton } from '@inera/ids-react'
import { useEffect, useState } from 'react'
import { Icon } from '../Icon/Icon'
import { classNames } from '../utils/classNames'

export function ScrollTopButton() {
  const [show, setShow] = useState(false)
  const [isScrolling, updateIsScrolling] = useState(false)

  useEffect(() => {
    let previousScrollPos = 0
    const handleScroll = () => {
      const scrollPos = document.body.getBoundingClientRect().top
      const isScrollInterupted = previousScrollPos > 0 && scrollPos > previousScrollPos

      setShow(scrollPos < -180)
      if (isScrolling && (scrollPos === 0 || isScrollInterupted)) {
        updateIsScrolling(false)
        window.open('#top', '_self')
        previousScrollPos = 0
      }
      previousScrollPos = scrollPos
    }

    window.addEventListener('scroll', handleScroll, true)
    return () => {
      window.removeEventListener('scroll', handleScroll, true)
    }
  }, [isScrolling])

  return (
    <div
      className={classNames(
        'sticky bottom-0 h-16 opacity-0 transition-opacity ease-in-out pointer-events-none',
        show && 'opacity-100 pointer-events-auto'
      )}
    >
      <IDSButton
        fab
        role="button"
        className="ids-hide ids-show-1177 absolute bottom-5 right-5 select-none"
        aria-label="Till toppen av sidan"
        onClick={() => {
          updateIsScrolling(true)
          window.scrollTo({
            top: 0,
            left: 0,
            behavior: 'smooth',
          })
        }}
      >
        <Icon icon="arrow-up" />
      </IDSButton>
    </div>
  )
}
