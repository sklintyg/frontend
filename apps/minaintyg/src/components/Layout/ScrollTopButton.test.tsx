import { render } from '@testing-library/react'
import { ScrollTopButton } from './ScrollTopButton'

it('Should render as expected', () => {
  const { baseElement } = render(<ScrollTopButton />)
  expect(baseElement).toMatchSnapshot()
})
