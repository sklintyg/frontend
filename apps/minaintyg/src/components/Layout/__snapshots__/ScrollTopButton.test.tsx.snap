// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render as expected 1`] = `
<body>
  <div>
    <div
      class="sticky bottom-0 h-16 opacity-0 transition-opacity ease-in-out"
    >
      <ids-button
        class="ids-hide ids-show-1177 absolute bottom-5 right-5 select-none"
      >
        <ids-icon-arrow
          class="ids-icon"
        />
      </ids-button>
    </div>
  </div>
</body>
`;
